# <center><u>Data Cleaning</u></center>

 Data cleaning is a part of the process on a data science project.
 
Data cleaning is the process of fixing or removing incorrect, corrupted, incorrectly formatted, duplicate, or incomplete data within a dataset.
<br>
When you clean your data, all outdated or incorrect information is gone – leaving you with the highest quality information for your analysis and model building.

In [1]:
# import pandas and numpy with alias pd and np respectively
import numpy as np
import pandas as pd

Create a dataframe, d1 = pd.DataFrame( {‘Temperature’ : [1, np.nan, 3, 2, 3] ,’Humidity’ : [22, np.nan, 2 , np.nan, 20 ] })

In [ ]:
#create d1
d1 = pd.DataFrame({'Temperature':[1,np.nan,3,2,3],'Humidity':[22,np.nan,2,np.nan,20]})

print the dataframe d1

In [ ]:
#print d1
d1

Temperature  Humidity
0          1.0      22.0
1          NaN       NaN
2          3.0       2.0
3          2.0       NaN
4          3.0      20.0

Find whether the given dataframe contain any missing values? 

In [ ]:
#check for null
d1.isnull()

Temperature  Humidity
0        False     False
1         True      True
2        False     False
3        False      True
4        False     False

How many missing values does each column have?

In [ ]:
#total null
d1.isnull().sum()

Temperature    1
Humidity       2
dtype: int64

## Dealing with missing values



<br>Now, we know we have missing values, the next thing that we need to work on, is how to deal with these missing values

 - ## Method 1: Delete the rows which contain missing values.
 This method include dropping all the rows that have missing value in any column. 

Use a suitable method to drop all the rows having missing values and save the change in d2 variable

In [ ]:
#drop nan
d2 = d1.dropna()

Print d2

In [ ]:
#print nan
d2

Temperature  Humidity
0          1.0      22.0
2          3.0       2.0
4          3.0      20.0

Hey Remember : droping rows with nan is one of the method to deal with missing values. But you have to decide if you need to go for this method by checking percentage of nan present in the dataframe.

If a column is having more than 60% of nan values then its better to remove such variables altogether if business permits 

- ## Method 2: Replacing missing values
Sometimes rather than dropping NA values, you'd rather replace them with a valid value. This value might be a single number like zero, or it might be some sort of imputation or interpolation

Impute the missing values with constant number of your choice

In [ ]:
# The below output has imputed missing  values with 100

d_nan = d1.fillna(100)
d_nan


Temperature  Humidity
0          1.0      22.0
1        100.0     100.0
2          3.0       2.0
3          2.0     100.0
4          3.0      20.0

Do you think its a  good way to treat Nan values? What if such constant values are not suitable for our further analysis?
Try to give your thoughts on this.

In [ ]:
"""No, treating Nan values with random number is not a great choice, it can make data skewed, or biased with one particular value"""

'No, treating Nan values with random number is not a great choice, it can make data skewed, or biased with one particular value'

Impute the missing values with mean

In [ ]:
#  imputing mean
d_mean = d1.fillna({'Temperature':np.mean(d1['Temperature']),'Humidity': np.mean(d1['Humidity'])})
d_mean

Temperature   Humidity
0         1.00  22.000000
1         2.25  14.666667
2         3.00   2.000000
3         2.00  14.666667
4         3.00  20.000000

Impute the missing values with median

In [ ]:
#median imputing

d_median = d1.fillna({'Temperature': d1['Temperature'].median(), 'Humidity':d1['Humidity'].median()})
d_median

Temperature  Humidity
0          1.0      22.0
1          2.5      20.0
2          3.0       2.0
3          2.0      20.0
4          3.0      20.0

Replacing with the mean, mode or median approximations are a statistical approach of handling the missing values.

Another Fun fact:
    
    This is an approximation which can add variance to the data set. But the loss of the data can be negated by this method which yields better results compared to removal of rows and columns.

Impute Nan with forward fill

In [ ]:
#forward fill
d_ffill = d1.fillna(method='ffill')
d_ffill

Temperature  Humidity
0          1.0      22.0
1          1.0      22.0
2          3.0       2.0
3          2.0       2.0
4          3.0      20.0

Impute Nan with backward fill

In [ ]:
#backward fill
d_bfill = d1.fillna(method='bfill')
d_bfill

Temperature  Humidity
0          1.0      22.0
1          3.0       2.0
2          3.0       2.0
3          2.0      20.0
4          3.0      20.0

Hey a fun fact here, as sweet as a cookie:

- ffill/pad/bfill are good imputation method if our data is of time series. This would keep the trend unaffected for our analysis.

Impute nan using interpolation method

In [ ]:
#interpolate
d_ipol = d1.interpolate()
d_ipol

Temperature  Humidity
0          1.0      22.0
1          2.0      12.0
2          3.0       2.0
3          2.0      11.0
4          3.0      20.0


 - Interpolation method by default is linear in nature. It is an imputation technique that assumes a linear relationship between data points and utilises non-missing values from adjacent data points to compute a value for a missing data point.




Perform KNN imputation


In [ ]:
from sklearn.impute import KNNImputer


In [ ]:
# Hint: Import KNNImputer and impute it on d1. Also note: Use n_neighbors=2
imputer = KNNImputer(n_neighbors=2)
d_knn = pd.DataFrame(imputer.fit_transform(d1), columns= d1.columns)
d_knn

Temperature   Humidity
0         1.00  22.000000
1         2.25  14.666667
2         3.00   2.000000
3         2.00  12.000000
4         3.00  20.000000

Point to ponder: KNN is an algorithm that is useful for matching a point with its closest k neighbors in a multi-dimensional space.

Do you think scaling is required to implement this method?. Yes you are right the answer is YES.
Can you comment below why normalized data is required, so that we understand your logic on this.

It requires us to normalize our data. Otherwise, the different scales of our data will lead the KNN Imputer to generate biased replacements for the missing values.

# Dropping Irrelevant Columns



Create a dataframe df = pd.DataFrame(np.random.randint(0,100,size=(100, 5)), columns=list('ABCDE'))


In [ ]:
np.random.seed(10)

df = pd.DataFrame(np.random.randint(0,100,size=(100,5)),columns=list('ABCDE'))


print df

In [ ]:
# print df
df

A   B   C   D   E
0    9  15  64  28  89
1   93  29   8  73   0
2   40  36  16  11  54
3   88  62  33  72  78
4   49  51  54  77  69
..  ..  ..  ..  ..  ..
95   3  50  59  34  21
96  16  18  61  54  60
97  21  87  83  71  16
98  67  38  27  96  87
99  98  89  16  82  19

[100 rows x 5 columns]

Note: Since all the rows are having random numbers, your dataframe observations might be different than the output given above

Suppose for our analysis our project do not require column E. So you need to remove this column. update this new change using inplace parameter

In [ ]:
#drop E
df.drop(columns='E', inplace= True)

Check if column **E** is removed by printing head of df

In [ ]:
#df head
df.head()

A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77

# Ensure requirements as per domain



Shallow copy the dataframe df in variable df2 and print df2 head






In [ ]:
df2= df.copy(deep= False)

#df2 head
df2.head()

A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77

Suppose your domain expert says to filter column B with even numbers to do correct analysis. Implement the same below and update the change in varaible df2.

In [ ]:
df2= df2[df2['B']%2==0]

print updated head of *df2*

In [ ]:
# df2 head
df2.head()

A   B   C   D
2   40  36  16  11
3   88  62  33  72
6   30  30  89  12
9   11  28  74  88
10  15  18  80  71

# Creating a sensible index values



Oops. The index in this dataframe doesn't make sense. please correct index in a sequential manner starting from 1. Save the updates in df2

In [ ]:
#set proper index
df2.index = [x for x in range(1, len(df2.values)+1)]

print df2 head again

In [ ]:
# df2 head
df2.head()

A   B   C   D
1  40  36  16  11
2  88  62  33  72
3  30  30  89  12
4  11  28  74  88
5  15  18  80  71

# Renaming column names to meaningful names.




Now df2 columns represents marks of the adventurous 'Anand', the brave 'Barkha', the compassionate 'Chandu' and the dashing 'Daniel'. Rename the columns with their name inplace of their first letter of their name.

In [ ]:
#column renaming
df2.rename(columns={'A':'Anand','B':'Barkha','C':'Chandu','D':'Daniel'}, inplace= True)

print df2 tail with updated column names

In [ ]:
# df2 head
df2.tail()

Anand  Barkha  Chandu  Daniel
44      1      82      34      11
45     74      36       6      63
46      3      50      59      34
47     16      18      61      54
48     67      38      27      96

Yeah! now the data looks pretty meaningful to study

# Treating Duplicate Data


Make another dataframe df3 by deep copying df2.

In [ ]:
df3= df2.copy(deep = True)

Make another column in df3 with name 'dummy' having 0 as values
throughout the rows.

In [ ]:
#assign dummy column full of zero value
df3['dummy'] = 0

In [ ]:
# print head of df, df2 and df3
print(df.head())
print(df2.head())
print(df3.head())


    A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77
   Anand  Barkha  Chandu  Daniel
1     40      36      16      11
2     88      62      33      72
3     30      30      89      12
4     11      28      74      88
5     15      18      80      71
   Anand  Barkha  Chandu  Daniel  dummy
1     40      36      16      11      0
2     88      62      33      72      0
3     30      30      89      12      0
4     11      28      74      88      0
5     15      18      80      71      0


Don't you think, there is some difference between copy operation used for creating df2 and df3.



In [ ]:
# copy provides two method shallow and deep, shaloow method actually points to the original dataset, so when we imply any change to the any of the dataset it will reflect back to both the dataset original and copied
# while when we use deep method, it creates another memory location for the copied dataframe, and any changes made to any one of the dataset will remain intact with themself only

print tail of df3

In [ ]:
# df3 tail
df3.tail()

Anand  Barkha  Chandu  Daniel  dummy
44      1      82      34      11      0
45     74      36       6      63      0
46      3      50      59      34      0
47     16      18      61      54      0
48     67      38      27      96      0

make an array name 'ListB' with values of column 'Barkha'

In [ ]:
ListB= np.array(df3['Barkha'])

Print ListB

In [ ]:
#print ListB
ListB

array([36, 62, 30, 28, 18, 50, 88, 50, 80, 66, 96, 30,  4, 30,  2, 42, 94,
       18, 44, 68, 58, 48, 70, 22, 36, 32, 32, 96, 30, 86,  0, 76, 88, 64,
       52, 46, 20, 66, 56,  8, 68, 50, 28, 82, 36, 50, 18, 38])

Assign this array values as another column in df3 with name 'Anonymous'

In [ ]:
#create Anonymous column
df3['Anonymous'] = pd.DataFrame(ListB,index= df3.index)

In [ ]:
df3.head()

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
1     40      36      16      11      0         36
2     88      62      33      72      0         62
3     30      30      89      12      0         30
4     11      28      74      88      0         28
5     15      18      80      71      0         18

#Create a dataframe 'ListA' with values of row index 3, 10 and 40 -->

In [ ]:
ListA= df3.loc[[3,10,40]]

print ListA

In [ ]:
# print ListA
ListA

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
3      30      30      89      12      0         30
10     96      66      67      62      0         66
40     74       8      92      32      0          8

Concat ListA to df3 ignoring the index values of ListA so that we can maintain the sequential index value thoughout the dataframe.

In [ ]:
df3= df3.append(ListA, ignore_index= True)

print head of df3

In [ ]:
# df3 head
df3.head()

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
0     40      36      16      11      0         36
1     88      62      33      72      0         62
2     30      30      89      12      0         30
3     11      28      74      88      0         28
4     15      18      80      71      0         18

Check if there is any duplicate rows present in the dataframe df3

In [ ]:
#check duplicate
df3[df3.duplicated()]

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
48     30      30      89      12      0         30
49     96      66      67      62      0         66
50     74       8      92      32      0          8

By above output it seems we do have duplicated rows in our dataset

Drop duplicated rows using pandas function keeping first values of such duplicated observations

In [ ]:
#drop duplicate
df3.drop_duplicates(keep='first',inplace= True)

Check again if we have any duplicate row values present 

In [ ]:
#check duplicate
def check_duplicate_columns(df):
  #create empty set
  duplicateColumnsName = set()
  #iterate through all column values
  for x in range(df.shape[1]):
    #take column at xth index
    col = df.iloc[:,x]
    #iterate through all columns from x+1 to last column
    for y in range(x+1, df.shape[1]):
      # take column at yth index
      otherCol = df.iloc[:, y]
      # check if two columns at xth and yth are eual or not
      # add to set if equal
      if col.equals(otherCol):
        duplicateColumnsName.add(df.columns.values[y])
    # Return list of unique column names 
    # whose contents are duplicates.
  return list(duplicateColumnsName)

 


In [ ]:
check_duplicate_columns(df3)

['Anonymous']

In [ ]:
# Get list of duplicate columns
duplicateColumnsName = check_duplicate_columns(df3)
  
print('Duplicate Columns are :')
        
    # Iterate through duplicate
    # column names
for column in duplicateColumnsName :
  print('Column Name : ', column)

Duplicate Columns are :
Column Name :  Anonymous


Yipeee!! Did you notice the dataframe is free from any duplicate rows now.

Drop any duplicate columns present in the dataframe df

In [ ]:
df3.head()

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
0     40      36      16      11      0         36
1     88      62      33      72      0         62
2     30      30      89      12      0         30
3     11      28      74      88      0         28
4     15      18      80      71      0         18

In [ ]:
df3 = df3.T.drop_duplicates().T

#print df3
df3

Anand  Barkha  Chandu  Daniel  dummy
0      40      36      16      11      0
1      88      62      33      72      0
2      30      30      89      12      0
3      11      28      74      88      0
4      15      18      80      71      0
5      88      50      54      34      0
6      77      88      15       6      0
7      97      50      45      40      0
8      81      80      41      90      0
9      96      66      67      62      0
10     88      96      73      40      0
11     28      30      89      25      0
12     33       4      87      94      0
13     68      30      70      74      0
14      9       2      65      13      0
15     62      42      34      40      0
16     32      94      86      58      0
17     45      18      50      44      0
18      6      44       9      50      0
19     44      68      14       4      0
20     39      58      81      39      0
21     69      48      60      58      0
22      5      70      61      16      0
23      4      22      19      85      0
24     10      36       0      30      0
25     85      32      81      13      0
26     31      32      96      38      0
27      0      96      25      63      0
28      2      30      17      59      0
29     63      86      29      71      0
30     19       0       7      93      0
31     58      76      10      79      0
32      7      88      96      38      0
33     27      64      49      90      0
34     27      52      90      57      0
35     99      46      46      45      0
36     84      20      49      33      0
37     77      66      37      31      0
38     82      56      41      90      0
39     74       8      92      32      0
40     87      68      86      14      0
41     50      50      85      14      0
42      1      28      58      10      0
43      1      82      34      11      0
44     74      36       6      63      0
45      3      50      59      34      0
46     16      18      61      54      0
47     67      38      27      96      0

Did you notice which Column is dropped?
I am sure you noticed it.
Name that column below

In [ ]:
#Column:_Anonymous___________

# Treating constant column values



Check unique values in each columns

In [ ]:
# df3 unique value
df3.nunique(axis =0)

Anand     42
Barkha    32
Chandu    40
Daniel    36
dummy      1
dtype: int64

By above output which column has only 1 value as unique throught the rows? 
Yeah! you are right, its dummy column.
So lets drop it

Drop dummy column as it has constant values which will not give us any information and save the changes to df3 using inplace parameter

In [ ]:
# drop dummy
df3 = df3.drop('dummy', axis= 1)

print final obtained dataframe df3

In [ ]:
# print df3
df3

Anand  Barkha  Chandu  Daniel
0      40      36      16      11
1      88      62      33      72
2      30      30      89      12
3      11      28      74      88
4      15      18      80      71
5      88      50      54      34
6      77      88      15       6
7      97      50      45      40
8      81      80      41      90
9      96      66      67      62
10     88      96      73      40
11     28      30      89      25
12     33       4      87      94
13     68      30      70      74
14      9       2      65      13
15     62      42      34      40
16     32      94      86      58
17     45      18      50      44
18      6      44       9      50
19     44      68      14       4
20     39      58      81      39
21     69      48      60      58
22      5      70      61      16
23      4      22      19      85
24     10      36       0      30
25     85      32      81      13
26     31      32      96      38
27      0      96      25      63
28      2      30      17      59
29     63      86      29      71
30     19       0       7      93
31     58      76      10      79
32      7      88      96      38
33     27      64      49      90
34     27      52      90      57
35     99      46      46      45
36     84      20      49      33
37     77      66      37      31
38     82      56      41      90
39     74       8      92      32
40     87      68      86      14
41     50      50      85      14
42      1      28      58      10
43      1      82      34      11
44     74      36       6      63
45      3      50      59      34
46     16      18      61      54
47     67      38      27      96

# Iterating dataframes

Let's look at three main ways to iterate over DataFrames.

1. iteritems()
2. iterrows()
3. itertuples()

We will also see time taken by these methods to print our dataframe. 

**1. Iterating DataFrames with iteritems()**

Lets iterate over rows of df3 uisng iteritems.



In [ ]:
import time
start = time.time()

#Use iteritems to iterate

for i,j in df3.iteritems():
  print(i)
  print(j)

print('Time taken(sec): ',(time.time()-start)*1000)

Anand
0     40
1     88
2     30
3     11
4     15
5     88
6     77
7     97
8     81
9     96
10    88
11    28
12    33
13    68
14     9
15    62
16    32
17    45
18     6
19    44
20    39
21    69
22     5
23     4
24    10
25    85
26    31
27     0
28     2
29    63
30    19
31    58
32     7
33    27
34    27
35    99
36    84
37    77
38    82
39    74
40    87
41    50
42     1
43     1
44    74
45     3
46    16
47    67
Name: Anand, dtype: int64
Barkha
0     36
1     62
2     30
3     28
4     18
5     50
6     88
7     50
8     80
9     66
10    96
11    30
12     4
13    30
14     2
15    42
16    94
17    18
18    44
19    68
20    58
21    48
22    70
23    22
24    36
25    32
26    32
27    96
28    30
29    86
30     0
31    76
32    88
33    64
34    52
35    46
36    20
37    66
38    56
39     8
40    68
41    50
42    28
43    82
44    36
45    50
46    18
47    38
Name: Barkha, dtype: int64
Chandu
0     16
1     33
2     89
3     74
4     80
5     54
6     15


Did you notice, how iteritems are iterating over df3.

Along with ways each iterating function works, also keep tallying the time taken for all other lopps too!. This will be fun, lets check iterrows()

**2. Iterating DataFrames with iterrows()**

In [ ]:
import time
start = time.time()
#Use iterrows to iterate
for i,j in df3.iterrows():
  print(i)
  print(j)

print('Time taken(sec): ',(time.time()-start)*1000)

0
Anand     40
Barkha    36
Chandu    16
Daniel    11
dummy      0
Name: 0, dtype: int64
1
Anand     88
Barkha    62
Chandu    33
Daniel    72
dummy      0
Name: 1, dtype: int64
2
Anand     30
Barkha    30
Chandu    89
Daniel    12
dummy      0
Name: 2, dtype: int64
3
Anand     11
Barkha    28
Chandu    74
Daniel    88
dummy      0
Name: 3, dtype: int64
4
Anand     15
Barkha    18
Chandu    80
Daniel    71
dummy      0
Name: 4, dtype: int64
5
Anand     88
Barkha    50
Chandu    54
Daniel    34
dummy      0
Name: 5, dtype: int64
6
Anand     77
Barkha    88
Chandu    15
Daniel     6
dummy      0
Name: 6, dtype: int64
7
Anand     97
Barkha    50
Chandu    45
Daniel    40
dummy      0
Name: 7, dtype: int64
8
Anand     81
Barkha    80
Chandu    41
Daniel    90
dummy      0
Name: 8, dtype: int64
9
Anand     96
Barkha    66
Chandu    67
Daniel    62
dummy      0
Name: 9, dtype: int64
10
Anand     88
Barkha    96
Chandu    73
Daniel    40
dummy      0
Name: 10, dtype: int64
11
Anand     28
Bar

**3. Iterating DataFrames with itertuples()**

In [ ]:
#iterate df3 using itertuples
import time
start = time.time()

#Use itertuples to iterate
for row in df3.itertuples():
  print(row)
  


print('Time taken(sec): ',(time.time()-start)*1000)

Pandas(Index=0, Anand=40, Barkha=36, Chandu=16, Daniel=11, dummy=0)
Pandas(Index=1, Anand=88, Barkha=62, Chandu=33, Daniel=72, dummy=0)
Pandas(Index=2, Anand=30, Barkha=30, Chandu=89, Daniel=12, dummy=0)
Pandas(Index=3, Anand=11, Barkha=28, Chandu=74, Daniel=88, dummy=0)
Pandas(Index=4, Anand=15, Barkha=18, Chandu=80, Daniel=71, dummy=0)
Pandas(Index=5, Anand=88, Barkha=50, Chandu=54, Daniel=34, dummy=0)
Pandas(Index=6, Anand=77, Barkha=88, Chandu=15, Daniel=6, dummy=0)
Pandas(Index=7, Anand=97, Barkha=50, Chandu=45, Daniel=40, dummy=0)
Pandas(Index=8, Anand=81, Barkha=80, Chandu=41, Daniel=90, dummy=0)
Pandas(Index=9, Anand=96, Barkha=66, Chandu=67, Daniel=62, dummy=0)
Pandas(Index=10, Anand=88, Barkha=96, Chandu=73, Daniel=40, dummy=0)
Pandas(Index=11, Anand=28, Barkha=30, Chandu=89, Daniel=25, dummy=0)
Pandas(Index=12, Anand=33, Barkha=4, Chandu=87, Daniel=94, dummy=0)
Pandas(Index=13, Anand=68, Barkha=30, Chandu=70, Daniel=74, dummy=0)
Pandas(Index=14, Anand=9, Barkha=2, Chandu=65,


    
    iteritems(): Helps to iterate over each element of the set, column-wise. 
    iterrows(): Each element of the set, row-wise. 
    itertuple(): Each row and form a tuple out of them.

But if you ask for speed. The most best perfromance is given by itertuples compared to other two iterating methods.
So if anytime you need to save your computation time on iteration of dataframes you can go for itertuples. Was'nt it fun?:)

# Regular Expression 


Python has a module named re to work with RegEx


### **!Are you ready to try regex on dataframes?**

*So here we go.!*

We are gonna try out following awesome re module functions

1. findall 
2. search
3. sub
4. split



In [3]:
from google.colab import files
uploaded = files.upload()

Saving trump_insult_tweets_2014_to_2021.csv to trump_insult_tweets_2014_to_2021.csv


In [44]:
import io
#load dataset
tweet_data= pd.read_csv(io.BytesIO(uploaded['trump_insult_tweets_2014_to_2021.csv']))

In [5]:
#import re module
import re

#print head of tweet_data
tweet_data.head()

Unnamed: 0  ...                                              tweet
0           1  ...  Can you believe this fool, Dr. Thomas Frieden ...
1           2  ...  Can you believe this fool, Dr. Thomas Frieden ...
2           3  ...  Big time in U.S. today - MAKE AMERICA GREAT AG...
3           4  ...  Politician @SenatorCardin didn't like that I s...
4           5  ...  For the nonbeliever, here is a photo of @Neily...

[5 rows x 5 columns]

Lets do some analysis using regex on this dataset


**1. findall()**

Make another column 'year' with year in each row using regex on date column.


In [45]:
# create a function which takes date as parameter and applies regex on it
def get_year(dd):
  pattern = r'([0-9]{4})\b'
  result = re.findall(pattern,dd)[0]
  return result

#use apply function on tweet_data to use above function in order to make year column
tweet_data['year'] = tweet_data['date'].apply(lambda dd: get_year(dd))

#print tweet_data head
tweet_data.head()

Unnamed: 0  ...  year
0           1  ...  2014
1           2  ...  2014
2           3  ...  2015
3           4  ...  2015
4           5  ...  2015

[5 rows x 6 columns]

lets filter year from 2020-2021 which was the election time in USA.

**2. search**

We will use regex search for this 

In [46]:
# create a function which takes year as parameter and applies regex on it
def ext_year(dt):
  x = re.search("^202\d", dt)
  if x:
    return dt
    
#use apply function on tweet_data to use above function in order to search tweets of 2020-2021
tweet_data['year'] = tweet_data['year'].apply(ext_year)

#tweet_data head
tweet_data.head()

Unnamed: 0  ...  year
0           1  ...  None
1           2  ...  None
2           3  ...  None
3           4  ...  None
4           5  ...  None

[5 rows x 6 columns]

In [ ]:
tweet_data['year'] = tweet_data['year'].str.match(pat = '(2020)|(2021)')
tweet_data

In [38]:
def ext_year(dd):
  result = dd.str.match(pat = (2020)|(2021))
  return result

In [47]:
tweet_data[tweet_data['year'].str.match('^202\d') == True]

Unnamed: 0  ...  year
7621         7622  ...  2020
7622         7623  ...  2020
7623         7624  ...  2020
7624         7625  ...  2020
7625         7626  ...  2020
...           ...  ...   ...
10355       10356  ...  2021
10356       10357  ...  2021
10357       10358  ...  2021
10358       10359  ...  2021
10359       10360  ...  2021

[2739 rows x 6 columns]

cool right!

You got some null values after applying above function. Lets drop them using dropna function. Also drop 'Unnamed: 0' column as it does not give any information.


In [48]:
#drop na and Unnamed: 0 column
tweet_data.dropna(inplace = True)
tweet_data.drop(['Unnamed: 0'], axis= 1, inplace= True)


**3. sub() Function**

Now you have filtered the dataset with 2739 rows. Let's remove all @ from tweet column suing re sub() function.

In [49]:
# create a function which takes tweet as parameter and applies regex on it
def tweets(tweet):
  pattern = r'@'
  result = re.sub(pattern,'', tweet)
  return result


#use apply function on tweet_data to use above function in order to remove @ from tweets
tweet_data['tweet'] = tweet_data['tweet'].apply(tweets)
#tweet_data head
tweet_data.head()

date  ...  year
7621  2020-01-01  ...  2020
7622  2020-01-02  ...  2020
7623  2020-01-04  ...  2020
7624  2020-01-06  ...  2020
7625  2020-01-06  ...  2020

[5 rows x 5 columns]

You can also use the sub function just in one line using list comprehension. Can you try doing it below?

In [50]:
# sub() suing list comprehension
tweet_data['tweet'] = [re.sub(r'@','',str(x)) for x in tweet_data['tweet']]

**4. split() Function**

Lets now split the column target by making another column Name which has name before hyphen("-")


In [51]:
# create a function which takes target as parameter and applies regex on it
def name(target):
  x = re.split('-', target, 1)
  if x:
    return x[0]

#use apply function on tweet_data to use above function in order to create Name column
tweet_data['Name'] = tweet_data['target'].apply(name)
#tweet_data head
tweet_data.head()

date               target  ...  year         Name
7621  2020-01-01  impeachment-inquiry  ...  2020  impeachment
7622  2020-01-02  impeachment-inquiry  ...  2020  impeachment
7623  2020-01-04  impeachment-inquiry  ...  2020  impeachment
7624  2020-01-06          los-angeles  ...  2020          los
7625  2020-01-06  impeachment-inquiry  ...  2020  impeachment

[5 rows x 6 columns]

Now can you filter out Name which are specially targetted for trump? Lets do it below and check how many such tweets are there.

In [52]:
#filter Name which is euqals to trump
tweet_data[tweet_data['Name'] == 'trump']

date         target  ...  year   Name
7859   2020-02-09   trump-russia  ...  2020  trump
7860   2020-02-09   trump-russia  ...  2020  trump
7861   2020-02-09   trump-russia  ...  2020  trump
7890   2020-02-12   trump-russia  ...  2020  trump
7891   2020-02-12   trump-russia  ...  2020  trump
...           ...            ...  ...   ...    ...
9658   2020-10-07   trump-russia  ...  2020  trump
9659   2020-10-07   trump-russia  ...  2020  trump
10091  2020-11-22   trump-russia  ...  2020  trump
10188  2020-12-09  trump-critics  ...  2020  trump
10258  2020-12-18   trump-russia  ...  2020  trump

[65 rows x 6 columns]

So here you got total of 65 records which were  tweeted on Donald Trump in the span of 2020-2021. 



Good job!  Now our interesting trump insult tweet data is somewhat cleaned. 